Refresh topic list view on update message events

kaustubh-nair opened the pull request.

kaustubh-nair commented on zulipterminal/model.py:
```diff
@@ -945,9 +945,20 @@ def _handle_update_message_event(self, event: Event) -> None:
                     self._update_rendered_view(msg_id)
+                if stream_id in self.index['topics']:
+                    # Reload topic list.
+                    self._fetch_topics_in_streams([stream_id])
```
@neiljp No, as I mentioned on czo, due to the topic data in the index never being complete, there is no way of knowing when a topic is empty (i.e doesn't have any messages left).

It would be useful if the backend response could include an extra flag when this happens, though it's not present currently AFAIK.

neiljp replied:
Let's discuss on #zulip-terminal.

neiljp commented:
@kaustubh-nair I think we've clarified via czo that it's OK to use the topic fetch approach, though as I mentioned elsewhere it'd be useful if we didn't fetch it unless/until we actively need it (for the subsequent conditional, or possibly another state we may be in?). For example, if we were in a stream narrow and triggered a topic rename and not showing the topic list, we could wait until autocompleting topics or showing the topic narrow before fetching it?

You didn't respond regarding the first point for the refactoring, but we can chat on czo regarding this if you like.

kaustubh-nair commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -183,12 +183,7 @@ def mark_unmuted(self, unread_count: int) -> None:
-                topic_view,
-                self.view.left_panel.options(height_type="weight")
-                )
+            self.view.left_panel.show_topics_view(self)
```
I've done this by making two methods, one each for updating topics view and stream view, and making minor tweaks in `update_stream_view` (previously `update_structure`). I hope this covers everything.

kaustubh-nair commented:
@neiljp Thanks for the review! I have made the topic fetch conditional, resetting the cache when not needed. I've also incorporated the refactor you mentioned (and a few others), if something has been left out in that let me know.

Tests fixed and added.

neiljp commented:
@kaustubh-nair As per comment in the stream just now, I've just merged most of the early commits with minor text/layout changes. The topic is not being updated in some cases, so I'd like to debug that before merging.
